### ISSUE TITLE
Fitting does not respect type parameters

### ISSUE THREAD

**Issue by gdalle** [JuliaStats/Distributions.jl#1544]
Hi!
In my current project, I would like to fit distributions in reduced precision to speed up my code. I was therefore wondering if the following behavior is intentional:
```julia
julia> using Distributions

julia> fit(Normal{Float32}, [1, 2, 3])
Normal{Float64}(μ=2.0, σ=0.816496580927726)
```

**Comment by devmotion** [comment]
It's known, the type only depends on the data. If you use `Float32[1, 2, 3]` the type should be `Normal{Float32}`.

**Comment by devmotion** [comment]
Of course, the question is still if this should be changed. I think it would be reasonable.

**Comment by gdalle** [comment]
Thanks for your answer. Indeed, when presented like this the current behavior seems logical too, since sufficient statistics are computed from the sample. But I agree that both options can make sense, and it is hard to decide which one should prevail

**Comment by gdalle** [comment]
Hum, after testing it doesn't go the way you promised it would :sweat_smile: 
```julia
julia> using Distributions

julia> fit(Normal{Float32}, Float32[1, 2, 3])
Normal{Float64}(μ=2.0, σ=0.816496580927726)
```

**Comment by devmotion** [comment]
Well, it seems there are some hardcoded Float64 arguments, e.g. in https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/continuous/normal.jl#L238. It would be easy to change these type restrictions at least in a first step.

**Comment by gdalle** [comment]
I think one would also need to make sufficient stats parametric, for instance here:
https://github.com/JuliaStats/Distributions.jl/blob/dd6ae8f4eac304f404b0069540a6c3bb1c667f92/src/univariate/continuous/normal.jl#L108-L113

**Comment by matbesancon** [comment]
yeah there are still a lot of F64 hardcoded sprinkled in the package :/

**Comment by timholy** [comment]
In case it helps resolve difficult design decisions,

```julia
help?> fit_mle
search: fit_mle

  fit_mle(D, x)

  Fit a distribution of type D to a given data set x.
```

From this, there's no doubt that the type parameters of `D` should take precedence over the eltype of the data. Either the docs need changing or the behavior should match the docs.

**Comment by gdalle** [comment]
I had given it a first shot months ago in this PR:

https://github.com/JuliaStats/Distributions.jl/pull/1560

Looking back at it, my main issue is that we need to accommodate both

- `fit(Normal, x)`, in which case the eltype of the data is the right choice
- `fit(Normal{Float64}, x)`, in which case the eltype of the distribution is the right choice

**Comment by timholy** [comment]
I think something like #1823 would make it an easy fix, as the final line of any code that builds distributions just becomes `D(args...)` and if the user specified the eltype of `D` then it will be respected.

**Comment by gdalle** [comment]
I think your attempt and mine fix different parts of the problem?
Even once we have internal coherence in the types and coercion thanks to #1823, we will still need to ensure that sufficient statistics are also type-generic, as well as all the code in between. We also have to enforce the conversion at the end of `fit`, which #1560 does
